### ISSUE TITLE
ndsctl strange behaviour in docker environment

### ISSUE THREAD

**Issue by gueux** [openNDS/openNDS#148]
Hi, I have successfully build and run openNDS inside docker containers on top of [BalenaOS](https://www.balena.io/os/)
It works very well, except ndsctl command line utility.

When I try to use it inside container as:

`# ndsctl -s /tmp/opennds/ndsctl.sock deauth 10.42.0.123`

I get the error:

`Client 10.42.0.123,(null),RESIN_APP_ID=1 not found.`

`RESIN_APP_ID=1` - this is one of built-in environment variable in docker container.

```
# printenv | grep RESIN_APP_ID
RESIN_APP_ID=1
```
So, it looks like at some point in the original attribute "10.42.0.123", some additional objects "(null),RESIN_APP_ID=1" had been added.

I've reviewed openDNS src by its still not clear for me how environment variables might be added to socket writer.
Could you assume how it might happened and how i can troubleshoot it?

**Comment by bluewavenet** [comment]
@gueux 
I have never used the -s option so just tried it here on OpenWrt - I can confirm it does not work!
I have also just tested with NoDogSplash 3.2.1 on an old OpenWrt system from 2018 and it does not work there either.
It probably has not worked for a very long time - if ever. It is not obvious what is wrong.
It is interesting that you are seeing it somehow picking up something from the local environment - this hints at some sort of memory overrun or similar.

However:
I would tend to class this as legacy code and deprecate it. All it does is allow you to move the socket file... Do we need it?
Obviously you have tried to use it. Any specific reason other than putting the socket file somewhere tidy?

**Comment by bluewavenet** [comment]
@gueux 
We could get rid of ndsctl all together and add its useful options to the openNDS command line. A simple ndsctl shim script could be provided for compatibility...

**Comment by bluewavenet** [comment]
@gueux 

> its still not clear for me how environment variables might be added to socket writer

It is a bit of code I have not paid any attention to before. It looks very cumbersome. Also it has not changed for a VERY long time so more or less confirming the buggy option is vary rarely used if ever... until you tried it ;-)

This brings me back to the question - Do we need socket access at all?  Why not simplify it all and use direct command line access to opennds?

Your opinions would be welcome.

**Comment by bluewavenet** [comment]
@gueux 
But of course direct command line access would bring its own difficulties eg I was thinking shared memory....
Maybe a rewrite of the socket code....

**Comment by gueux** [comment]
@bluewavenet Originally I require socket file in non default place because of docker environment.
My setup looks like this:
```
services:
  dynamic-fas-nodejs:
    volumes:
      - "ndsctl:/tmp/opennds/"
  captive-portal-opennds:
    network_mode: host
    volumes:
      - "ndsctl:/tmp/opennds/"
  volumes:
    ndsctl: {}
 ```

My goal is to have isolated containers for openNDS and FAS service (NodeJS Express.js app), but also have ability to interact between two applications. As there is no any API in openNDS, its possible to do through socket shared through docker volume only. So, my node.js app just use socket for auth/deauth users onto openNDS side. And generally i don't need ndsctl utility for my goal, I just use raw socket through [net::Socket](https://nodejs.org/api/net.html#net_class_net_socket), but some of raw socket calls doesn't work correctly also. As example `deauth ip` works well, but `auth ip` crashed opennds with 'segmentation fault' error.
Just let me know if i can help you with any additional troubleshooting.

**Comment by bluewavenet** [comment]
@gueux 
I don't follow what you are trying to do, but yes the current socket implementation has problems.
Apart from bugs, it is also slow. I am finding a single ndsctl call can take over 1.5 seconds for a completed reply. It is high on my list of things to fix!

> As there is no[t] an[] API in openNDS

FAS and openNDS use a REST API to communicate with each other, provided at the openNDS end by libmicrohttpd (the authmon daemon helps with inbound NAT traversal).
Authmon does in fact use ndsctl, but could just as well use raw socket access... It has been in my thoughts for a while to look at using shared memory access for this when time permits.

> auth ip crashed opennds with 'segmentation fault' error

What version of openNDS are you using?

**Comment by bluewavenet** [comment]
@gueux 
I found the problem. It occurs only when the ndsctl -s option is used and is caused by incorrect parsing of the passed command line arguments. This is fixed in the upcoming v9.0.0 release.
If you need a code snippet I can post it here.

**Comment by bluewavenet** [comment]
@gueux 
OpenNDS 9.0.0 is available now.
(It was merged with OpenWrt 21.02 and master on 14th May 2021. It is also compatible with 19.07 as long as you upgrade libmicrohttp as well)
